walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,70 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Comparator;
+import java.util.Arrays;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+      Collection<Event> eventsAL = new ArrayList<Event>(events);
+      Collection<String> people = request.getAttendees();
+
+      Collection<TimeRange> availableTimes = new ArrayList<TimeRange>();
+      availableTimes.add(TimeRange.WHOLE_DAY);
+      getAvailableTimes(eventsAL, people, availableTimes, request.getDuration());
+    
+      Collection<TimeRange> availableTimesOptional = new ArrayList<TimeRange>(availableTimes);
+      Collection<String> optionalPeople = request.getOptionalAttendees();
+      getAvailableTimes(eventsAL, optionalPeople, availableTimesOptional, request.getDuration());
+
+      if(request.getAttendees().isEmpty() && availableTimesOptional.isEmpty()){
+          return Arrays.asList();
+      }
+
+      return availableTimesOptional.isEmpty() ? availableTimes : availableTimesOptional;
+  }
+
+  private void getAvailableTimes(Collection<Event> events, Collection<String> people, Collection<TimeRange> availableTimes, long duration){
+      Iterator<Event> it = events.iterator();
+      while(it.hasNext()){
+          Event element = it.next();
+          if(!Collections.disjoint(element.getAttendees(), people)){
+            createAvailableRange(availableTimes, element.getWhen());
+            it.remove();
+          }
+      }
+      availableTimes.removeIf(item -> (item.duration() < duration));
+      return;
+  }
+
+  private void createAvailableRange(Collection<TimeRange> availableTimes, TimeRange newEvent){
+      List<TimeRange> deleteList = new ArrayList<TimeRange>();
+      for(TimeRange event : availableTimes){
```

Comment:
The code checks `availableTimes` multiple times. This might be unnecessary: if we have a 5pm meeting, then we don't need to consider an earlier free time slot ending at 8am. Consider sorting all of the events by start time. Then the end time of the busy event that we're considering is the start of the next possible free slot.

Reply:
only checks once now